File: immersinglinker-sdk-py/immersinglinker/clients/automation.py
from __future__ import annotations

import json
from uuid import uuid4

import aiohttp

from ..enums import RuleSetSatisfyMode
from ..types import (
    AutomationPlan,
    AutomationPlanInfo,
    Trigger,
    Action,
    RuleSet,
    RuleBase,
    TriggerDto,
    ActionDto,
    RuleNodeDto,
    RuleSetDto,
    CreateAutomationPlanRequest,
    UpdateAutomationPlanRequest,
)
from .base import ImmersingLinkerError


class AutomationServiceClient:
    def __init__(self, port: str) -> None:
        self._base_url = f"http://localhost:{port}"
        self._session: aiohttp.ClientSession | None = None

    async def __aenter__(self) -> AutomationServiceClient:
        self._session = aiohttp.ClientSession(self._base_url)
        return self

    async def __aexit__(self, *args: object) -> None:
        if self._session:
            await self._session.close()
            self._session = None

    async def _ensure_session(self) -> aiohttp.ClientSession:
        if self._session is None:
            raise ImmersingLinkerError(
                "Client not opened. Use 'async with AutomationServiceClient(...) as client:'"
            )
        return self._session

    async def _request(
        self, method: str, path: str, body: object = None, *, raise_on_404: bool = False
    ) -> str:
        session = await self._ensure_session()
        kwargs: dict = {}
        if body is not None:
            kwargs["json"] = self._serialize(body)
        async with session.request(method, path, **kwargs) as response:
            text = await response.text()
            if response.status == 404:
                if raise_on_404:
                    raise ImmersingLinkerError(f"Not found: {text}")
                return ""
            if response.status == 400:
                raise ImmersingLinkerError(f"Bad request: {text}")
            if not 200 <= response.status < 300:
                raise ImmersingLinkerError(f"HTTP {response.status}: {text}")
            return text

    @staticmethod
    def _serialize(obj: object) -> object:
        if obj is None:
            return None
        if isinstance(obj, (str, int, float, bool)):
            return obj
        if isinstance(obj, dict):
            return {k: AutomationServiceClient._serialize(v) for k, v in obj.items()}
        if isinstance(obj, (list, tuple)):
            return [AutomationServiceClient._serialize(v) for v in obj]
        if hasattr(obj, "to_dict"):
            return obj.to_dict()  # type: ignore[union-attr]
        if hasattr(obj, "__dataclass_fields__"):
            from dataclasses import asdict

            return asdict(obj)  # type: ignore[call-overload]
        return str(obj)

    # region GET

    async def get_all_plan_infos(self) -> list[AutomationPlanInfo]:
        text = await self._request("GET", "/automation")
        data = json.loads(text) if text else []
        return [AutomationPlanInfo.from_dict(item) for item in data]

    async def get_plan_by_guid(self, plan_guid: str) -> AutomationPlan | None:
        text = await self._request("GET", f"/automation/{plan_guid}")
        return AutomationPlan.from_dict(json.loads(text)) if text else None

    # endregion

    # region POST

    async def create_plan(self, request: CreateAutomationPlanRequest) -> AutomationPlan:
        text = await self._request("POST", "/automation", body=request)
        return AutomationPlan.from_dict(json.loads(text))

    async def trigger_plan(self, plan_guid: str) -> None:
        await self._request("POST", f"/automation/{plan_guid}/trigger", raise_on_404=True)

    async def invoke_url_trigger(self, tag: str) -> None:
        await self._request("POST", f"/automation/invoke/{tag}")

    # endregion

    # region PUT

    async def update_plan(
        self, plan_guid: str, request: UpdateAutomationPlanRequest
    ) -> AutomationPlan:
        text = await self._request("PUT", f"/automation/{plan_guid}", body=request, raise_on_404=True)
        return AutomationPlan.from_dict(json.loads(text))

    # endregion

    # region DELETE

    async def delete_plan(self, plan_guid: str) -> None:
        await self._request("DELETE", f"/automation/{plan_guid}", raise_on_404=True)

    # endregion

    # region Offline factories

    @staticmethod
    def create_automation_plan_offline(
        name: str,
        revertable: bool,
        trigger: Trigger,
        rule_set: RuleSet | None,
        actions: list[Action],
    ) -> AutomationPlan:
        return AutomationPlan(
            Guid=uuid4(),
            Name=name,
            Revertable=revertable,
            Trigger=trigger,
            RuleSet=rule_set
            or RuleSet(SatisfyMode=RuleSetSatisfyMode.AllSatisfied, Not=False),
            Actions=actions,
        )

    @staticmethod
    def create_trigger_dto_offline(
        trigger_key: str, properties: dict | None = None
    ) -> TriggerDto:
        return TriggerDto(TriggerKey=trigger_key, Properties=properties)

    @staticmethod
    def create_rule_set_offline(
        satisfy_mode: RuleSetSatisfyMode, not_: bool, rules: list[RuleNodeDto]
    ) -> RuleSetDto:
        return RuleSetDto(SatisfyMode=satisfy_mode, Not=not_, Rules=rules)

    @staticmethod
    def create_rule_node_offline(
        rule_key: str | None = None,
        properties: dict | None = None,
        not_: bool = False,
        rule_set: RuleSetDto | None = None,
    ) -> RuleNodeDto:
        return RuleNodeDto(
            RuleKey=rule_key, Properties=properties, Not=not_, RuleSet=rule_set
        )

    @staticmethod
    def create_action_dto_offline(
        action_key: str, properties: dict | None = None
    ) -> ActionDto:
        return ActionDto(ActionKey=action_key, Properties=properties)

    @staticmethod
    def create_plan_request_offline(
        name: str,
        revertable: bool,
        trigger: TriggerDto,
        rule_set: RuleSetDto | None,
        actions: list[ActionDto],
    ) -> CreateAutomationPlanRequest:
        return CreateAutomationPlanRequest(
            Name=name,
            Revertable=revertable,
            Trigger=trigger,
            RuleSet=rule_set,
            Actions=actions,
        )

    @staticmethod
    def update_plan_request_offline(
        name: str,
        revertable: bool,
        trigger: TriggerDto,
        rule_set: RuleSetDto | None,
        actions: list[ActionDto],
    ) -> UpdateAutomationPlanRequest:
        return UpdateAutomationPlanRequest(
            Name=name,
            Revertable=revertable,
            Trigger=trigger,
            RuleSet=rule_set,
            Actions=actions,
        )

    # endregion

    async def close(self) -> None:
        if self._session:
            await self._session.close()
            self._session = None

File: immersinglinker-sdk-node/src/clients/automation.ts
import { ServiceClientBase } from './base.js';
import type {
  AutomationPlan,
  AutomationPlanInfo,
  CreateAutomationPlanRequest,
  UpdateAutomationPlanRequest,
  TriggerDto,
  ActionDto,
  RuleSetDto,
  RuleNodeDto,
  RuleSetSatisfyMode,
} from '../types/automation.js';
import { v4 as uuidv4 } from 'uuid';

/** 自动化管理客户端，提供自动化方案的 CRUD 及触发操作 */
export class AutomationServiceClient extends ServiceClientBase {
  // #region GET

  /** 获取所有自动化方案的摘要信息 */
  async getAllPlanInfos(): Promise<AutomationPlanInfo[]> {
    return this._get<AutomationPlanInfo[]>('/automation');
  }

  /** 根据 GUID 获取指定自动化方案，不存在时返回 null */
  async getPlanByGuid(planGuid: string): Promise<AutomationPlan | null> {
    return this._getOrNull<AutomationPlan>(`/automation/${planGuid}`);
  }

  // #endregion

  // #region POST

  /** 创建自动化方案 */
  async createPlan(
    request: CreateAutomationPlanRequest,
  ): Promise<AutomationPlan> {
    return this._post<AutomationPlan>('/automation', request);
  }

  /** 手动触发指定自动化方案 */
  async triggerPlan(planGuid: string): Promise<void> {
    return this._postVoid(`/automation/${planGuid}/trigger`);
  }

  /** 通过标签触发 UrlTrigger */
  async invokeUrlTrigger(tag: string): Promise<void> {
    return this._postVoid(`/automation/invoke/${tag}`);
  }

  // #endregion

  // #region PUT

  /** 更新指定自动化方案 */
  async updatePlan(
    planGuid: string,
    request: UpdateAutomationPlanRequest,
  ): Promise<AutomationPlan> {
    return this._put<AutomationPlan>(`/automation/${planGuid}`, request);
  }

  // #endregion

  // #region DELETE

  /** 删除指定自动化方案 */
  async deletePlan(planGuid: string): Promise<void> {
    return this._delete(`/automation/${planGuid}`);
  }

  // #endregion

  // #region Offline factory methods

  /** 创建一个离线触发器 DTO */
  static createTriggerDtoOffline(
    triggerKey: string,
    properties?: unknown,
  ): TriggerDto {
    return { triggerKey, properties };
  }

  /** 创建一个离线规则集 DTO */
  static createRuleSetDtoOffline(
    satisfyMode: RuleSetSatisfyMode,
    not: boolean,
    rules: RuleNodeDto[],
  ): RuleSetDto {
    return { satisfyMode, not, rules };
  }

  /** 创建一个离线规则节点 DTO */
  static createRuleNodeDtoOffline(
    ruleKey: string | null,
    properties: unknown | undefined,
    not: boolean,
    ruleSet: RuleSetDto | null,
  ): RuleNodeDto {
    return {
      ruleKey,
      properties,
      not,
      ruleSet,
    };
  }

  /** 创建一个离线动作 DTO */
  static createActionDtoOffline(
    actionKey: string,
    properties?: unknown,
  ): ActionDto {
    return { actionKey, properties };
  }

  /** 创建一个离线创建方案的请求对象 */
  static createPlanRequestOffline(
    name: string,
    revertable: boolean,
    trigger: TriggerDto,
    ruleSet: RuleSetDto | null,
    actions: ActionDto[],
  ): CreateAutomationPlanRequest {
    return { name, revertable, trigger, ruleSet, actions };
  }

  /** 创建一个离线更新方案的请求对象 */
  static updatePlanRequestOffline(
    name: string,
    revertable: boolean,
    trigger: TriggerDto,
    ruleSet: RuleSetDto | null,
    actions: ActionDto[],
  ): UpdateAutomationPlanRequest {
    return { name, revertable, trigger, ruleSet, actions };
  }

  // #endregion
}